File: erlang/mathtree.erl
-module(mathtree).

-export([parse/1]).

parse([String]) ->
    run(string:tokens(String, " ")).

run(Tokens) ->
    {_, _, Start} = now(),
    Expressions = make_expr(make_terms(Tokens, []), []),
    io:fwrite("~p~n", [Expressions]),
    Result = calculate(Expressions),
    {_, _, End} = now(),
    Time = End - Start,
    io:fwrite("time = ~w result = ~w~n", [Time, Result]),
    init:stop().

%%
%% make_terms/2
%% build a list of terms from the list of tokens
%% finally turn those terms into expressions
%%
make_terms([], Out) -> lists:reverse(Out);
make_terms(["+"|T], Out) -> make_terms(T, [add|Out]);
make_terms(["-"|T], Out) -> make_terms(T, [subtract|Out]);
make_terms(["*"|T], Out) -> make_terms(T, [multiply|Out]);
make_terms(["/"|T], Out) -> make_terms(T, [divide|Out]);
make_terms([Number|T], Out) -> make_terms(T, [list_to_integer(Number)|Out]).


%%
%% make_expr/2
%% terms are either a number or an atom denoting an operation
%% create nested expressions in the form of
%% {operator, operand-1, operand-2}
%%
make_expr([Term|Terms], Numbers) when is_number(Term) ->
    make_expr(Terms, [Term|Numbers]);
make_expr([Term|Terms], [N1, N2|Numbers]) -> %% it is an atom
    case Terms of
	[] -> {Term, N2, N1};
	_ ->
	    make_expr(Terms, [{Term, N2, N1}|Numbers])
	end.

%%
%% calculate the expression
%%
calculate(N) when is_number(N) ->
    N;
calculate({add, Op1, Op2}) ->
    calculate(Op1) + calculate(Op2);
calculate({subtract, Op1, Op2}) ->
    calculate(Op1) - calculate(Op2);
calculate({multiply, Op1, Op2}) ->
    calculate(Op1) * calculate(Op2);
calculate({divide, Op1, Op2}) ->
    calculate(Op1) div calculate(Op2).

File: erlang/mathtree.rb
OPERATOR = "operator"
OP1 = "op1"
OP2 = "op2"

ADD = "add"
SUBTRACT = "subtract"
MULTIPLY = "multiply"
DIVIDE = "divide"

def calculate(data)
  if data.is_a?(Numeric)
    data
  elsif data.is_a?(Hash)
    if data[OPERATOR] == ADD
	    calculate(data[OP1]) + calculate(data[OP2])
    elsif data[OPERATOR] == SUBTRACT
	    calculate(data[OP1]) - calculate(data[OP2])
    elsif data[OPERATOR] == MULTIPLY
	    calculate(data[OP1]) * calculate(data[OP2])
    elsif data[OPERATOR] == DIVIDE
	    calculate(data[OP1]) / calculate(data[OP2])
    end
  else
    puts "how did I get here" + data.class.name
  end
end

def make_terms(input, output)
  if input.length == 0
    return output
  else
    element = input.shift
    number = element.to_i
    if element.match(/[0-9]+/)
	    output << element.to_i
    elsif element == "+"
	    output << ADD
    elsif element == "-"
	    output << SUBTRACT
    elsif element == "x"
	    output << MULTIPLY
    elsif element == "/"
	    output << DIVIDE
    end
    make_terms(input, output)
  end
end

def make_expr(terms, expression)
  top = terms.shift
  if top
    if top.is_a?(Numeric)
      # puts "number: " + top.to_s + " " + expression.to_s
	    expression << top
    else
      # puts "term: " + top + " " + expression.to_s
	    expression << { OPERATOR => top, OP2 => expression.pop(), OP1 => expression.pop() }
    end
    make_expr(terms, expression)
  else
    # showtree(expression)
    expression.last
  end
end

0.upto(999) {
  instring = Array.new(ARGV)
  start_time = Time.now.usec
  data = make_expr(make_terms(instring, Array.new()), Array.new())
  # puts "Expressions: " + data.join(" ")
  result = calculate(data)
  end_time = Time.now.usec
  puts "Ruby\ttime = " + (end_time - start_time).to_s + "\tresult = " + result.to_s
}
